feat(4D-ROI): ROI Segmentation panel (4D)

lscoder opened the pull request:
- Updated `ToolGroupServices` to make it possible to add tool instances (eg: circular brush/eraser tools are instances of BrushTool)
- Updated `calculateSUVPeak` to support 4D datasets since volumes can have multiple arrays of `scalarData`
- Added ROI Segmentation panel to 4D mode (started with a copy of that same panel from TMTV)
- Updated ROI Segmentation panel layout and created a `BrushConfiguration` component to make it possible to change all brush-related tools settings
- Added all segmentation tools (circular brush, circular eraser, sphere brush, sphere eraser, threshold brush, rectangle ROI threshold, rectangle scissor, circle scissor, sphere scissor and paint fill)
- Updated the toolbar to keep all segmentation related tools in a single group
- Updated `thresholdSegmentationByRectangleROITool` and `calculatedSUVPeak` from TMTV to also support `Rectangle ROI Threshold` (only `Rectangle ROI Threshold (start/end)` was supported before)
- Created `ToolService` to be able to trigger and listen to events (activated tools)
- Fixed 4D mode HP (PT viewports were using a wrong `toolGroup`) + a few other changes to make it similar to HP from TMTV
- Fixed some issues in `ViewportWindowLevel` related to segmentations

lscoder commented on extensions/cornerstone-dynamic-volume/src/getHangingProtocolModule.ts:
```diff
@@ -300,29 +298,10 @@ const defaultProtocol = {
       },
     },
   ],
-  toolGroupIds: ['dynamic4D-default'],
   // -1 would be used to indicate active only, whereas other values are
   // the number of required priors referenced - so 0 means active with
   // 0 or more priors.
-  numberOfPriorsReferenced: 0,
-  // Default viewport is used to define the viewport when
-  // additional viewports are added using the layout tool
-  defaultViewport: {
-    viewportOptions: {
-      viewportType: 'volume',
-      toolGroupId: 'dynamic4D-default',
-      allowUnmatchedView: true,
-      initialImageOptions: {
-        preset: 'middle', // 'first', 'last', 'middle'
-      },
-    },
-    displaySets: [
-      {
-        id: 'defaultDisplaySetId',
-        matchedDisplaySetsIndex: -1,
-      },
-    ],
-  },
```
Just trying to make it more similar to TMTV

lscoder replied:
Just trying to make it more similar to TMTV

sedghi commented on modes/preclinical-4d/src/index.tsx:
```diff
@@ -101,39 +110,39 @@ function modeFactory({ modeConfiguration }) {
           name: 'Registration',
           hangingProtocol: {
-            protocolId: 'default4D',
+            protocolId: preclinical4d.hangingProtocol,
```
I thought in each step we are applying a new protocol no?

sedghi replied:
hmmm this is confusing then and requires the a preexisting knowledge of the hanging protcol and that it should have multiple stages defined. Explicit code is much better in this situation so at least put a stage index or something as the argument

lscoder replied:
I'm not sure if I'm following because all workflow stages are using the same `protocolId` but each one of them has a different `stageId`.

sedghi replied:
Nevermind I understand now. Also when did we choose to add `workflow` to the mode root key? I thought we are using it in the onModeEnter and register the steps via the workflowService no?

Also seems like we have stages in the HP and stages in workflow which is a lot confusing we should think about how to rename the workflow stage to something else. I think we should opt in for `WorkflowStep` and `WorkflowStepService`

lscoder replied:
Renamed "stages" to "steps" everywhere

sedghi commented on extensions/cornerstone-dynamic-volume/src/getHangingProtocolModule.ts:
```diff
@@ -280,7 +277,8 @@ function getFusionViewports() {
 
 const defaultProtocol = {
-  id: 'default4D',
+  id:
```
i don't think you should make the full pathname here, if you say 'default' it automatically should create the  `@ohif/extension-cornerstone-dynamic-volume.hangingProtocolModule.default` if i'm not mistaken

lscoder replied:
I changed the id when I was comparing it with TMTV HP and I saw that one using `@ohif/extension-tmtv.hangingProtocolModule.ptCT` as id

sedghi replied:
does it work if you say `default` here? if so please do so

lscoder replied:
It works if we set it to default because it would use the default HP from default mode but I've reverted my change renaming the HP back to `default4D`

sedghi commented on extensions/cornerstone/src/services/ToolGroupService/ToolGroupService.ts:
```diff
@@ -316,16 +288,17 @@ export default class ToolGroupService {
-      // }
-
-      toolGroup.addTool(toolName, { ...toolConfig });
+    ['active', 'passive', 'enabled', 'disabled'].forEach(toolStateName => {
```
can we use enums for these?

lscoder replied:
Updated to use the Enum from `cornerstoneTools`

sedghi commented on extensions/cornerstone/src/init.tsx:
```diff
@@ -76,18 +76,13 @@ export default async function init({
     stateSyncService,
   } = servicesManager.services as CornerstoneServices;
 
-  window.services = servicesManager.services;
```
please don't remove them they are here for debugging the demos

lscoder replied:
I always search for `window.`, `debug` and `console.` when I run `git diff` before pushing and I deleted those lines by accident

sedghi commented on extensions/cornerstone/src/services/ToolService/ToolService.ts:
```diff
@@ -0,0 +1,152 @@
+    }
+  };
+
+  public getActivePrimaryMouseButtonTool(toolGroupId?: string): string {
+    const { toolGroupService } = this.servicesManager
+      .services as CornerstoneServices;
+
+    const toolGroup = toolGroupService.getToolGroup(toolGroupId);
+
+    return toolGroup?.getActivePrimaryMouseButtonTool();
+  }
```
this is something that I think toolGroupService should handle not toolService

lscoder replied:
Moved to `toolGroupService`

sedghi commented on extensions/cornerstone/src/services/ToolService/ToolService.ts:
```diff
@@ -0,0 +1,152 @@
+    'event::cornerstone::toolservice:primaryToolActivated',
+};
+
+class ToolService extends PubSubService {
```
I don't understand why we need another service. It appears that you are only subscribing to an event and just publishing again. You can do that in the init.ts for the cornerstone extension

sedghi replied:
if you really need a service I recommend you using `toolGroupService` for this

lscoder replied:
Updated

sedghi reviewed (commented):
thanks, see my comments

lscoder commented:
- Renamed workflow "stages" to "steps" everywhere
- ROI Segmentation is available only when user gets to "ROI Quantification" workflow step
- Fixed issues when updating left/right panels content based on workflow settings or moving across different tabs because components were losing their state (broke WorkflowSteps dropdown, WindowLevels)
- Fixed "Rectangle ROI Threshold (start/end)" which was not working on fusion viewports
- Removed "Threshold" setting from circular and sphere brush tools (only brush size is available)
- "Rectangle ROI Threshold" now have settings similar to "Rectangle ROI Threshold (start/end)" but without start/end buttons
